pkg/statdb/statdb.go

```diff
@@ -184,34 +194,67 @@ func (s *Server) Update(ctx context.Context, updateReq *pb.UpdateRequest) (resp
 	}, nil
 }
 
-// UpdateBatch for updating  multiple farmers' stats in the db
+// UpdateBatch for updating multiple farmers' stats in the db
 func (s *Server) UpdateBatch(ctx context.Context, updateBatchReq *pb.UpdateBatchRequest) (resp *pb.UpdateBatchResponse, err error) {
-	// todo(moby) how should we handle one node failing to update but not all?
 	defer mon.Task()(&ctx)(&err)
 	s.logger.Debug("entering statdb UpdateBatch")
 
 	APIKeyBytes := updateBatchReq.APIKey
-	nodeStatsList := make([]*pb.NodeStats, len(updateBatchReq.NodeList))
-	for i, node := range updateBatchReq.NodeList {
+	var nodeStatsList []*pb.NodeStats
+	var failedNodes []*pb.Node
+	for _, node := range updateBatchReq.NodeList {
 		updateReq := &pb.UpdateRequest{
 			Node:   node,
 			APIKey: APIKeyBytes,
 		}
 
 		updateRes, err := s.Update(ctx, updateReq)
 		if err != nil {
-			return nil, status.Errorf(codes.Internal, err.Error())
+			s.logger.Error(err.Error())
+			failedNodes = append(failedNodes, node)
+		} else {
+			nodeStatsList = append(nodeStatsList, updateRes.Stats)
 		}
-
-		nodeStatsList[i] = updateRes.Stats
 	}
 
 	updateBatchRes := &pb.UpdateBatchResponse{
-		StatsList: nodeStatsList,
+		FailedNodes: failedNodes,
+		StatsList:   nodeStatsList,
 	}
 	return updateBatchRes, nil
 }
 
+// CreateEntryIfNotExists creates a statdb node entry and saves to statdb if it didn't already exist
+func (s *Server) CreateEntryIfNotExists(ctx context.Context, createIfReq *pb.CreateEntryIfNotExistsRequest) (resp *pb.CreateEntryIfNotExistsResponse, err error) {
+	APIKeyBytes := createIfReq.APIKey
+	getReq := &pb.GetRequest{
+		NodeId: createIfReq.Node.NodeId,
+		APIKey: APIKeyBytes,
+	}
+	getRes, err := s.Get(ctx, getReq)
+	if err != nil {
+		if strings.Contains(err.Error(), "no rows in result set") {
```

Comment:
Is that dbx error check out of the question here? It would probably be better than string matching, and I thought you had a check for that before

Reply:
I tried, the check didn't hold up in testing.